arch/arm/arm64/include/uk/asm/paging.h

```diff
@@ -47,19 +65,213 @@ struct ukarch_pagetable {
+#define ARM64_INVALID_ADDR		0xBAADBAADBAADBAADUL
+
+#define __VADDR_INV					\
+	PAGE_Lx_ALIGN_DOWN(ARM64_INVALID_ADDR, PT_MAP_LEVEL_MAX)
+
+#define __PADDR_INV					\
+	PAGE_Lx_ALIGN_DOWN(ARM64_INVALID_ADDR, PT_MAP_LEVEL_MAX)
 
-/* TODO: Complement missing definitions */
+/* Any PTE with bit[0] == 0 is invalid */
+#define PT_Lx_PTE_INVALID(lvl)		0x0UL
+
+#define ARM64_PADDR_MAX ((1ULL << ARM64_PADDR_BITS) - 1)
+#define ARM64_VADDR_MAX ((1ULL << ARM64_VADDR_BITS) - 1)
+
+#define ARM64_PADDR_VALID(paddr) (paddr <= (__paddr_t)ARM64_PADDR_MAX)
+#define ARM64_VADDR_VALID(vaddr) (vaddr <= (__vaddr_t)ARM64_VADDR_MAX)
+
+#ifndef __ASSEMBLY__
+static inline __paddr_t PT_Lx_PTE_PADDR(__pte_t pte, unsigned int lvl)
+{
+	__paddr_t paddr;
+	__u64 pte_lx_map_paddr_mask[] = {
+		PTE_L0_PAGE_PADDR_MASK,
+		PTE_L1_BLOCK_PADDR_MASK,
+		PTE_L2_BLOCK_PADDR_MASK
+	};
+
+	if (PAGE_Lx_IS(pte, lvl)) {
+#ifdef CONFIG_LIBUKDEBUG
+		UK_ASSERT(lvl <= PT_MAP_LEVEL_MAX);
+#endif /* CONFIG_LIBUKDEBUG */
+		paddr = pte & pte_lx_map_paddr_mask[lvl];
+	} else {
+#ifdef CONFIG_LIBUKDEBUG
+		UK_ASSERT(lvl > PAGE_LEVEL && lvl < PT_LEVELS);
+#endif /* CONFIG_LIBUKDEBUG */
+		paddr = pte & PTE_Lx_TABLE_PADDR_MASK;
+	}
+	return paddr;
+}
+
+static inline int ukarch_paddr_range_isvalid(__paddr_t start, __paddr_t end)
+{
+#ifdef CONFIG_LIBUKDEBUG
+	UK_ASSERT(start <= end);
+#endif /* CONFIG_LIBUKDEBUG */
+
+	return (ARM64_PADDR_VALID(end)) && (ARM64_PADDR_VALID(start));
+}
+
+static inline int ukarch_vaddr_range_isvalid(__vaddr_t start, __vaddr_t end)
+{
+#ifdef CONFIG_LIBUKDEBUG
+	UK_ASSERT(start <= end);
+#endif /* CONFIG_LIBUKDEBUG */
+
+	return (ARM64_VADDR_VALID(end)) && (ARM64_VADDR_VALID(start));
+}
+
+static inline int ukarch_pte_read(__vaddr_t pt_vaddr, unsigned int lvl,
+				  unsigned int idx, __pte_t *pte)
+{
+	(void)lvl;
+
+#ifdef CONFIG_LIBUKDEBUG
+	UK_ASSERT(idx < PT_Lx_PTES(lvl));
+#endif /* CONFIG_LIBUKDEBUG */
+
+	*pte = *((__pte_t *)pt_vaddr + idx);
+
+	return 0;
+}
+
+static inline int ukarch_pte_write(__vaddr_t pt_vaddr, unsigned int lvl,
+				   unsigned int idx, __pte_t pte)
+{
+	(void)lvl;
+
+#ifdef CONFIG_LIBUKDEBUG
+	UK_ASSERT(idx < PT_Lx_PTES(lvl));
+#endif /* CONFIG_LIBUKDEBUG */
+
+	*((__pte_t *)pt_vaddr + idx) = pte;
+
+	return 0;
+}
+
+static inline __paddr_t ukarch_pt_read_base(void)
+{
+	__paddr_t reg;
+
+	__asm__ __volatile__("mrs %x0, ttbr0_el1\n" : "=r" (reg));
+
+	return (reg & TTBR0_EL1_BADDR_MASK);
+}
+
+static inline int ukarch_pt_write_base(__paddr_t pt_paddr)
+{
+	__paddr_t reg = (pt_paddr & TTBR0_EL1_BADDR_MASK);
+
+	__asm__ __volatile__("msr ttbr0_el1, %x0\n"
+			     "isb\n"
+			     :: "r" (reg));
+	return 0;
+}
+
+static inline void ukarch_tlb_flush_entry(__vaddr_t vaddr)
+{
+	__asm__ __volatile__(
+		"	dsb ishst\n"         /* wait for writes to complete */
+		"	tlbi vaae1is, %x0\n" /* invalidate by vaddr, any ASID */
+		"	dsb ish\n"           /* wait for invalidate complete */
+		"	isb\n"               /* sync context */
```

Comment:
I don't quite get the indentation of the ASM instructions here (and in `ukarch_tlb_flush`). Did you mean this?
```suggestion
		"dsb	ishst\n"         /* wait for writes to complete */
		"tlbi	vaae1is, %x0\n"  /* invalidate by vaddr, any ASID */
		"dsb	ish\n"           /* wait for invalidate complete */
		"isb\n"                  /* sync context */
```

Reply:
Another candidate for the coding style document 🙂 Previously we used syntax with prepended tabs, eg https://github.com/unikraft/unikraft/blob/staging/arch/x86/x86_64/include/uk/asm/spinlock.h#L76 so I assumed that this is the convention. Or does it only apply if we use labels? Btw, there are places where we use `asm volatile` and `__asm__ __volatile__`.  Also, if we choose not to prepend with tabs, we need to append them like `\n\t`.

Reply:
🤦🏻‍♂️ Oh yes sure, now I get it! There is no convention, if this has to be indented without labels. I would probably only do it if there are labels, but you are right with the `\n\t`. I never thought about that this way. So it makes sense to do it always 👍🏻 

I think the tab between instr. and operands makes also sense for multiline ASM listings to align operands as we also do in the `.S` files and the ASM code in the spinlock header. What do you think?

For the `__asm__` vs `asm`, I always go with `__asm__` because I have stored in the back of my head that it is more compatible. But I have not researched if this is actually true.

(agree, something that should go into the convention doc. I've put a note there so we don't forget)

Reply:
As before, no strong opinion. I only noticed now that prepending with tab + tab between arguments (makes sense) + comments after each instruction (also good practice) brings us often to beyond 80 chars.That requires to be creative with comments. Will update and use some newspeak to reduce comment size.

Reply:
Probably comments will get so encoded and short that only true ASM hackers understand 😎